{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import alphashape\n",
    "from descartes import PolygonPatch\n",
    "from shapely.geometry import Point\n",
    "from os import path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes in data matrix containing info for cell type to-be-curated and alphashape of structure in which \n",
    "# the cell type must be removed\n",
    "# Returns filtered cell type data\n",
    "def celltype_excluder(dat,alpha_shape):    \n",
    "    points = np.array(dat[['x','y']])\n",
    "    is_in_struct_arr = []\n",
    "    for coord in points:\n",
    "        is_in_struct = alpha_shape.intersects(Point(coord))\n",
    "        is_in_struct_arr.append(is_in_struct)\n",
    "\n",
    "    not_in_struct = [not i for i in is_in_struct_arr]\n",
    "    result = dat[not_in_struct]\n",
    "    \n",
    "    return(result)\n",
    "\n",
    "# Takes in data matrix containing info for cell type to-be-curated and alphashape of structure in which\n",
    "# the cell must be maintained\n",
    "# Returns filtered cell type data\n",
    "def celltype_includer(dat,alpha_shape):\n",
    "    points = np.array(dat[['x','y']])\n",
    "    is_in_struct_arr = []\n",
    "    for coord in points:\n",
    "        is_in_struct = alpha_shape.intersects(Point(coord))\n",
    "        is_in_struct_arr.append(is_in_struct)\n",
    "    \n",
    "    result = dat[is_in_struct_arr]\n",
    "    \n",
    "    return(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes in unique array id and outputs df with coordinates corresponding to all spatial inlier beads\n",
    "def get_inlier_coords(array_id):\n",
    "    # input_path is path to file with beads x features for all beads (spatial outliers removed)\n",
    "    # featuers = {'barcode','x','y'}\n",
    "    input_path = '{array_id}_coords.csv'.format(array_id=array_id)\n",
    "    all_coords = pd.read_csv(input_path,index_col=0)\n",
    "    return(all_coords)\n",
    "\n",
    "# Takes in unique array id and outputs df with raw cell type labels corresponding to all beads\n",
    "def get_allcells_unfiltered(array_id):\n",
    "    # input_path is path to file with beads x features for all beads (not yet curated)\n",
    "    # features = {'barcode','x','y','cell_type'}\n",
    "    input_path = '{array_id}_allcells_unfiltered_df.csv'.format(array_id=array_id)\n",
    "    allcells_unfiltered_df = pd.read_csv(input_path,index_col=0)\n",
    "    return(allcells_unfiltered_df)\n",
    "\n",
    "# Takes in unique array id and outputs df with raw cell type labels corresponding to all curated beads\n",
    "def get_allcells_filtered(array_id):\n",
    "    # input_path is path to file with beads x features for all curated beads\n",
    "    # features = {'barcode','x','y','cell_type'}\n",
    "    input_path = '{array_id}_allcells.csv'.format(array_id=array_id)\n",
    "    allcells_df = pd.read_csv(input_path,index_col=0)\n",
    "    return(allcells_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maps array id to genotype\n",
    "# (WT refers to BTBR wt/wt, DKD refers to BTBR ob/ob)\n",
    "d = {\n",
    "    '191204_15': 'WT',\n",
    "    '191204_22': 'DKD',\n",
    "    '191223_19': 'UMOD-WT',\n",
    "    '200104_07': 'UMOD-KI',\n",
    "    '200113_11': '21_cortex',\n",
    "    '200121_01': '21_medulla',\n",
    "    '200104_19': '12_cortex',\n",
    "    '200104_21': '12_medulla',\n",
    "    '200115_15': '19_cortex',\n",
    "    '200115_18': '19_medulla'\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blank arrays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    print(array_id)\n",
    "    geno = d[array_id]\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    \n",
    "    out_path = '{geno}_{array_id}_blank.pdf'.format(array_id=array_id,geno=geno)\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "    plt.xlim(500,5800)\n",
    "    plt.ylim(500,5800)\n",
    "    plt.axis('off')\n",
    "    plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close('all')\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Individually-curated cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    print(array_id)\n",
    "    \n",
    "    # input_path is path to file with beads x features for all beads assigned a curated cell type\n",
    "    # features = {'barcode','x','y','cell_type'}\n",
    "    input_path = '{array_id}_allcells_df.csv'.format(array_id=array_id)\n",
    "    dat = pd.read_csv(input_path,index_col=0)\n",
    "    cell_types = np.unique(dat['cell_type'])\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    \n",
    "    for cell_type in cell_types:\n",
    "        out_path = '{geno}_{array_id}_{cell_type}_curated_only.pdf'.format(geno=geno,array_id=array_id,cell_type=cell_type)\n",
    "        celltype_dat = dat[dat['cell_type']==cell_type].copy()\n",
    "        plt.figure(figsize=(5,5))\n",
    "        plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "        plt.scatter(celltype_dat['x'],celltype_dat['y'],s=10,c='r',ec='black',linewidth=0.3)\n",
    "        plt.xlim(500,5800)\n",
    "        plt.ylim(500,5800)\n",
    "        plt.axis('off')\n",
    "        plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close('all')\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curated vs. raw cell type plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate polygons encapsulating spatial inlier coordinates for every array\n",
    "# used later to ensure that unfiltered cell type calls are shown only within array area\n",
    "alphashape_d = {}\n",
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    \n",
    "    points = np.array(all_coords[['x','y']])\n",
    "    alpha = 0.01\n",
    "    alpha_shape = alphashape.alphashape(points,alpha=alpha)\n",
    "    \n",
    "    alphashape_d[array_id] = alpha_shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CD-IC detection in CD-PC tubules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    cell_type = 'CD-IC'\n",
    "        \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    allcells_unfiltered = get_allcells_unfiltered(array_id)\n",
    "    allcells_filtered = get_allcells_filtered(array_id)\n",
    "    \n",
    "    alpha_shape = alphashape_d[array_id]\n",
    "    \n",
    "    alpha = 0.01\n",
    "    has_glom = False\n",
    "    has_gc = False\n",
    "    has_md = False\n",
    "    \n",
    "    # input_path is path to file with beads x features for all beads in glomerulus\n",
    "    # features = {'barcode','x','y','cluster','cell_type'}\n",
    "    input_path = '{array_id}_glom_with_seurat_cells_info.csv'.format(array_id=array_id)\n",
    "    if path.exists(input_path):\n",
    "        has_glom = True\n",
    "        glom_df = pd.read_csv(input_path,index_col=0)\n",
    "        points = np.array(glom_df[['x','y']])\n",
    "        alpha_shape_glom = alphashape.alphashape(points,alpha=alpha)\n",
    "\n",
    "    gc_df = allcells_filtered[allcells_filtered['cell_type']=='Ren1'].copy()\n",
    "    if not gc_df.empty:\n",
    "        has_gc = True\n",
    "        points = np.array(gc_df[['x','y']])\n",
    "        alpha_shape_gc = alphashape.alphashape(points,alpha=alpha)\n",
    "\n",
    "    md_df = allcells_filtered[allcells_filtered['cell_type']=='MD'].copy()\n",
    "    if not md_df.empty:\n",
    "        has_md = True\n",
    "        points = np.array(md_df[['x','y']])\n",
    "        alpha_shape_md = alphashape.alphashape(points,alpha=alpha)\n",
    "    \n",
    "    celltype_unfiltered = allcells_unfiltered[allcells_unfiltered['cell_type']==cell_type].copy()\n",
    "    celltype_unfiltered = celltype_unfiltered[['barcode']].merge(all_coords[['barcode','x','y']],on='barcode')\n",
    "    celltype_unfiltered = celltype_includer(celltype_unfiltered,alpha_shape)\n",
    "\n",
    "    celltype_filtered = allcells_filtered[allcells_filtered['cell_type']==cell_type].copy()\n",
    "    celltype_filtered = celltype_filtered[['barcode','x','y']].copy()\n",
    "\n",
    "    cdpc_filtered = allcells_filtered[allcells_filtered['cell_type']=='CD-PC'].copy()\n",
    "    cdpc_filtered = cdpc_filtered[['barcode','x','y']].copy()\n",
    "\n",
    "    if has_glom:\n",
    "        celltype_unfiltered = celltype_excluder(celltype_unfiltered,alpha_shape_glom)\n",
    "    if has_gc:\n",
    "        celltype_unfiltered = celltype_excluder(celltype_unfiltered,alpha_shape_gc)\n",
    "    if has_md:\n",
    "        celltype_unfiltered = celltype_excluder(celltype_unfiltered,alpha_shape_md)\n",
    "\n",
    "    out_path = '{geno}_{array_id}_{cell_type}_raw_curated.pdf'.format(geno=geno,array_id=array_id,cell_type=cell_type)\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "    plt.scatter(celltype_unfiltered['x'],celltype_unfiltered['y'],s=10,c='dimgrey',ec='black',linewidth=0.3)\n",
    "    plt.scatter(cdpc_filtered['x'],cdpc_filtered['y'],s=10,c='yellowgreen',ec='black',linewidth=0.3)\n",
    "    plt.scatter(celltype_filtered['x'],celltype_filtered['y'],s=10,c='r',ec='black',linewidth=0.3)\n",
    "    plt.xlim(500,5800)\n",
    "    plt.ylim(500,5800)\n",
    "    plt.axis('off')\n",
    "    plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DCT, CD-PC, PCT detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    \n",
    "    if geno in ['DKD','WT','UMOD-KI','UMOD-WT']:\n",
    "        specimen = 'mouse'\n",
    "    else:\n",
    "        specimen = 'human'\n",
    "    \n",
    "    if specimen == 'mouse':\n",
    "        cell_types = ['DCT','CD-PC','PCT_1','PCT_2']\n",
    "    elif specimen == 'human':\n",
    "        cell_types = ['DCT','CD-PC','PCT']\n",
    "        \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    allcells_unfiltered = get_allcells_unfiltered(array_id)\n",
    "    allcells_filtered = get_allcells_filtered(array_id)\n",
    "    \n",
    "    alpha_shape = alphashape_d[array_id]\n",
    "    \n",
    "    alpha = 0.01\n",
    "    has_glom = False\n",
    "    has_gc = False\n",
    "    has_md = False\n",
    "    \n",
    "    # input_path is path to file with beads x features for all beads in glomerulus\n",
    "    # features = {'barcode','x','y','cluster','cell_type'}\n",
    "    input_path = '{array_id}_glom_with_seurat_cells_info.csv'.format(array_id=array_id)\n",
    "    if path.exists(input_path):\n",
    "        has_glom = True\n",
    "        glom_df = pd.read_csv(input_path,index_col=0)\n",
    "        glom_df = glom_df.drop(columns={'cluster'})\n",
    "        points = np.array(glom_df[['x','y']])\n",
    "        alpha_shape_glom = alphashape.alphashape(points,alpha=alpha)\n",
    "\n",
    "    gc_df = allcells_filtered[allcells_filtered['cell_type']=='Ren1'].copy()\n",
    "    if not gc_df.empty:\n",
    "        has_gc = True\n",
    "        points = np.array(gc_df[['x','y']])\n",
    "        alpha_shape_gc = alphashape.alphashape(points,alpha=alpha)\n",
    "\n",
    "    md_df = allcells_filtered[allcells_filtered['cell_type']=='MD'].copy()\n",
    "    if not md_df.empty:\n",
    "        has_md = True\n",
    "        points = np.array(md_df[['x','y']])\n",
    "        alpha_shape_md = alphashape.alphashape(points,alpha=alpha)\n",
    "    \n",
    "    for cell_type in cell_types:\n",
    "        print(puckid,pheno,cell_type)\n",
    "\n",
    "        celltype_unfiltered = allcells_unfiltered[allcells_unfiltered['max_celltype']==cell_type].copy()\n",
    "        celltype_unfiltered = celltype_unfiltered[['barcode']].merge(all_coords[['barcode','x','y']],on='barcode')\n",
    "        celltype_unfiltered = celltype_includer(celltype_unfiltered,alpha_shape)\n",
    "\n",
    "        celltype_filtered = allcells_filtered[allcells_filtered['cell_type']==cell_type].copy()\n",
    "        celltype_filtered = celltype_filtered[['barcode','x','y']].copy()\n",
    "        \n",
    "        if has_glom:\n",
    "            celltype_unfiltered = celltype_excluder(celltype_unfiltered,alpha_shape_glom)\n",
    "        if has_gc:\n",
    "            celltype_unfiltered = celltype_excluder(celltype_unfiltered,alpha_shape_gc)\n",
    "        if has_md:\n",
    "            celltype_unfiltered = celltype_excluder(celltype_unfiltered,alpha_shape_md)\n",
    "        \n",
    "        out_path = '{geno}_{array_id}_{cell_type}_raw_curated.pdf'.format(geno=geno,array_id=array_id,cell_type=cell_type)\n",
    "        plt.figure(figsize=(5,5))\n",
    "        plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "        plt.scatter(celltype_unfiltered['x'],celltype_unfiltered['y'],s=10,c='dimgrey',ec='black',linewidth=0.3)\n",
    "        plt.scatter(celltype_filtered['x'],celltype_filtered['y'],s=10,c='r',ec='black',linewidth=0.3)\n",
    "        plt.xlim(500,5800)\n",
    "        plt.ylim(500,5800)\n",
    "        plt.axis('off')\n",
    "        plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Podocyte detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    \n",
    "    if geno in ['DKD','WT','UMOD-KI','UMOD-WT']:\n",
    "        specimen = 'mouse'\n",
    "    else:\n",
    "        specimen = 'human'\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    allcells_unfiltered = get_allcells_unfiltered(array_id)\n",
    "    allcells_filtered = get_allcells_filtered(array_id)\n",
    "    \n",
    "    alpha_shape = alphashape_d[array_id]\n",
    "    \n",
    "    # input path is path to file containing beads x features for all beads in array\n",
    "    # (thresholded cell loading matrix output by NMFreg)\n",
    "    # features = {'barcode','x','y','max_cell_type','cell_type_1',...,'cell_type_n'} features = {'barcode','x','y','max_celltype','cell_type_1',...,'cell_type_n','maxval','thresh_ct'} (loading of each cell type across all beads, max loading per bead, cell type associated with max loading)\n",
    "    input_path = '{}_nmf_loadings.csv'.format(array_id)\n",
    "    nmf_calls = pd.read_csv(input_path,index_col=0)\n",
    "    if specimen == 'mouse':\n",
    "        celltype_unfiltered = nmf_calls[nmf_calls['max_cell_type']==10].copy()\n",
    "    elif specimen == 'human':\n",
    "        celltype_unfiltered = nmf_calls[nmf_calls['max_cell_type']==17].copy()\n",
    "    celltype_unfiltered = celltype_unfiltered[['barcode']].merge(all_coords[['barcode','x','y']],on='barcode')\n",
    "    celltype_unfiltered = celltype_includer(celltype_unfiltered,alpha_shape)\n",
    "\n",
    "    celltype_filtered = allcells_df[allcells_df['cell_type']=='Podocyte'].copy()\n",
    "    celltype_filtered = celltype_filtered[['barcode','x','y']].copy()\n",
    "\n",
    "    out_path = '{geno}_{array_id}_Podocyte_raw_curated.pdf'.format(geno=geno,array_id=array_id)\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "    plt.scatter(celltype_unfiltered['x'],celltype_unfiltered['y'],s=10,c='dimgrey',ec='black',linewidth=0.3)\n",
    "    plt.scatter(celltype_filtered['x'],celltype_filtered['y'],s=10,c='r',ec='black',linewidth=0.3)\n",
    "    plt.xlim(500,5800)\n",
    "    plt.ylim(500,5800)\n",
    "    plt.axis('off')\n",
    "    plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assigning non-podo cell types to gloms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    \n",
    "    if geno in ['DKD','WT','UMOD-KI','UMOD-WT']:\n",
    "        specimen = 'mouse'\n",
    "    else:\n",
    "        specimen = 'human'\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    \n",
    "    # input_path is path to file with beads x features for all beads in glomerulus\n",
    "    # features = {'barcode','x','y','cluster','cell_type'}\n",
    "    input_path = '{array_id}_glom_with_seurat_cells_info.csv'.format(array_id=array_id)\n",
    "    if path.exists(input_path):\n",
    "        glom_info = pd.read_csv(input_path,index_col=0)\n",
    "        if specimen == 'mouse':\n",
    "            d2 = {10:'Podocyte',12:'MC',1:'EC'}\n",
    "            new = [d2[x] for x in glom_info['max_cell_type']]\n",
    "            glom_info['max_cell_type'] = new\n",
    "        elif specimen == 'human':\n",
    "            d2 = {17:'Podocyte',16:'MC',7:'EC'}\n",
    "            new = [d2[x] for x in glom_info['max_cell_type']]\n",
    "            glom_info['max_cell_type'] = new\n",
    "        ec_info = glom_info[glom_info['max_cell_type']=='EC'].copy()\n",
    "        podo_info = glom_info[glom_info['max_cell_type']=='Podocyte'].copy()\n",
    "        mc_info = glom_info[glom_info['max_cell_type']=='MC'].copy()\n",
    "\n",
    "        out_path = '{geno}_{array_id}_assign_celltype_to_glom.pdf'.format(geno=geno,array_id=array_id)\n",
    "        plt.figure(figsize=(5,5))\n",
    "        plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "        plt.scatter(ec_info['x'],ec_info['y'],s=20,c='dodgerblue',ec='black',linewidth=0.6)\n",
    "        plt.scatter(podo_info['x'],podo_info['y'],s=20,c='midnightblue',ec='black',linewidth=0.6)\n",
    "        plt.scatter(mc_info['x'],mc_info['y'],s=20,c='cyan',ec='black',linewidth=0.6)\n",
    "        plt.xlim(500,5800)\n",
    "        plt.ylim(500,5800)\n",
    "        plt.axis('off')\n",
    "        plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GC detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    \n",
    "    if geno in ['DKD','WT','UMOD-KI','UMOD-WT']:\n",
    "        specimen = 'mouse'\n",
    "    else:\n",
    "        specimen = 'human'\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    allcells_unfiltered = get_allcells_unfiltered(array_id)\n",
    "    allcells_filtered = get_allcells_filtered(array_id)\n",
    "    \n",
    "    alpha_shape = alphashape_d[array_id]\n",
    "    \n",
    "    # input_path is path to gene expression counts matrix \n",
    "    input_path = '{array_id}_counts.csv'.format(array_id=array_id)\n",
    "    counts = pd.read_csv(input_path,index_col=0)\n",
    "    if specimen == 'mouse':\n",
    "        celltype_unfiltered = counts[counts['Ren1']>0].copy()\n",
    "    elif specimen == 'human':\n",
    "        celltype_unfiltered = counts[counts['REN']>0].copy()\n",
    "    celltype_unfiltered = celltype_unfiltered[['barcode']].merge(all_coords[['barcode','x','y']],on='barcode')\n",
    "    celltype_unfiltered = celltype_includer(celltype_unfiltered,alpha_shape)   \n",
    "    \n",
    "    glom_df = pd.DataFrame(columns=['barcode','x','y'])\n",
    "    # input_path is path to file with beads x features for all beads in glomerulus\n",
    "    # features = {'barcode','x','y','cluster','cell_type'}\n",
    "    input_path = '{array_id}_glom_with_seurat_cells_info.csv'.format(array_id=array_id)\n",
    "    if path.exists(input_path):\n",
    "        glom_df = pd.read_csv(input_path,index_col=0)\n",
    "        glom_df = glom_df[['barcode','x','y']].copy()\n",
    "\n",
    "    celltype_filtered = allcells_df[allcells_df['cell_type']=='Ren1'].copy()\n",
    "    celltype_filtered = celltype_filtered[['barcode','x','y']].copy()\n",
    "\n",
    "    out_path = '{geno}_{array_id}_GC_raw_curated.pdf'.format(geno=geno,array_id=array_id)\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "    plt.scatter(celltype_unfiltered['x'],celltype_unfiltered['y'],s=10,c='dimgrey',ec='black',linewidth=0.3)\n",
    "    plt.scatter(celltype_filtered['x'],celltype_filtered['y'],s=10,c='r',ec='black',linewidth=0.3)\n",
    "    plt.scatter(glom_df['x'],glom_df['y'],s=10,c='dodgerblue',ec='black',linewidth=0.3)\n",
    "    plt.xlim(500,5800)\n",
    "    plt.ylim(500,5800)\n",
    "    plt.axis('off')\n",
    "    plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MD detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for array_id in list(d.keys()):\n",
    "    geno = d[array_id]\n",
    "    print(array_id,geno)\n",
    "    \n",
    "    all_coords = get_inlier_coords(array_id)\n",
    "    allcells_unfiltered = get_allcells_unfiltered(array_id)\n",
    "    allcells_filtered = get_allcells_filtered(array_id)\n",
    "    \n",
    "    alpha_shape = alphashape_d[array_id]\n",
    "\n",
    "    celltype_unfiltered = allcells_unfiltered[allcells_unfiltered['max_celltype']=='TAL'].copy()\n",
    "    celltype_unfiltered = celltype_unfiltered[['barcode']].merge(all_coords[['barcode','x','y']],on='barcode')\n",
    "    celltype_unfiltered = celltype_includer(celltype_unfiltered,alpha_shape)\n",
    "    \n",
    "    celltype_filtered2 = allcells_df[allcells_df['cell_type']=='MD'].copy()\n",
    "    celltype_filtered2 = celltype_filtered2[['barcode','x','y']].copy()\n",
    "    \n",
    "    glom_df = pd.DataFrame(columns=['barcode','x','y'])\n",
    "    # input_path is path to file with beads x features for all beads in glomerulus\n",
    "    # features = {'barcode','x','y','cluster','cell_type'}\n",
    "    input_path = '{array_id}_glom_with_seurat_cells_info.csv'.format(array_id=array_id)\n",
    "    if path.exists(input_path):\n",
    "        glom_df = pd.read_csv(input_path,index_col=0)\n",
    "        glom_df = glom_df[['barcode','x','y']].copy()\n",
    "    \n",
    "    gc_df = allcells_df[allcells_df['cell_type']=='Ren1'].copy()\n",
    "    gc_df = gc_df[['barcode','x','y']].copy()\n",
    "    \n",
    "    out_path = '{geno}_{array_id}_MD_raw_curated.pdf'.format(geno=geno,array_id=array_id)\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(all_coords['x'],all_coords['y'],s=3,c='lightgrey')\n",
    "    plt.scatter(celltype_unfiltered['x'],celltype_unfiltered['y'],s=10,c='dimgrey',ec='black',linewidth=0.3)\n",
    "    plt.scatter(celltype_filtered2['x'],celltype_filtered2['y'],s=10,c='r',ec='black',linewidth=0.3)\n",
    "    plt.scatter(gc_df['x'],gc_df['y'],s=10,c='orange',ec='black',linewidth=0.3)\n",
    "    plt.scatter(glom_df['x'],glom_df['y'],s=10,c='dodgerblue',ec='black',linewidth=0.3)\n",
    "    plt.xlim(500,5800)\n",
    "    plt.ylim(500,5800)\n",
    "    plt.axis('off')\n",
    "    plt.savefig(out_path,dpi=300,bbox_inches='tight')\n",
    "    plt.show()\n",
    "    plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_legend(legend, filename=\"legend.pdf\"):\n",
    "    fig  = legend.figure\n",
    "    fig.canvas.draw()\n",
    "    bbox  = legend.get_window_extent().transformed(fig.dpi_scale_trans.inverted())\n",
    "    fig.savefig(filename, dpi=300, bbox_inches=bbox)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "plt.rcParams['font.size'] = 20\n",
    "plt.gca().spines['right'].set_color('none')\n",
    "plt.gca().spines['top'].set_color('none')\n",
    "plt.gca().spines['bottom'].set_color('none')\n",
    "plt.gca().spines['left'].set_color('none')\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "colors = ['dodgerblue',\"cyan\",'midnightblue']\n",
    "texts = ['EC','MC','Podocyte']\n",
    "patches = [ plt.plot([],[], marker=\"o\", ms=10, mec='black',ls='none', linewidth=0.6,color=colors[i], \n",
    "                label=\"{:s}\".format(texts[i]) )[0]  for i in range(len(texts)) ]\n",
    "legend2=plt.legend(handles=patches, loc=3, framealpha=1,frameon=False)\n",
    "export_legend(legend2,filename='legend1.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "plt.rcParams['font.size'] = 20\n",
    "plt.gca().spines['right'].set_color('none')\n",
    "plt.gca().spines['top'].set_color('none')\n",
    "plt.gca().spines['bottom'].set_color('none')\n",
    "plt.gca().spines['left'].set_color('none')\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "colors = ['dimgrey','r','dodgerblue','orange']\n",
    "texts = ['Raw TAL','Curated MD','Glomerulus','GC']\n",
    "patches = [ plt.plot([],[], marker=\"o\", ms=10, mec='black',ls='none', linewidth=0.6,color=colors[i], \n",
    "                label=\"{:s}\".format(texts[i]) )[0]  for i in range(len(texts)) ]\n",
    "legend2=plt.legend(handles=patches, loc=3, framealpha=1,frameon=False)\n",
    "export_legend(legend2,filename='legend2.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "plt.rcParams['font.size'] = 20\n",
    "plt.gca().spines['right'].set_color('none')\n",
    "plt.gca().spines['top'].set_color('none')\n",
    "plt.gca().spines['bottom'].set_color('none')\n",
    "plt.gca().spines['left'].set_color('none')\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "colors = ['dimgrey','r','dodgerblue']\n",
    "texts = ['Ren1/REN+','Curated GC','Glomerulus']\n",
    "patches = [ plt.plot([],[], marker=\"o\", ms=10, mec='black',ls='none', linewidth=0.6,color=colors[i], \n",
    "                label=\"{:s}\".format(texts[i]) )[0]  for i in range(len(texts)) ]\n",
    "legend2=plt.legend(handles=patches, loc=3, framealpha=1,frameon=False)\n",
    "export_legend(legend2,filename='legend3.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
